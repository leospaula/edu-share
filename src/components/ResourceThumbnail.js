import React from 'react'
import { Link } from 'react-router-dom'
import { displayGrade } from '../utilities'

const ResourceThumbnail = props => {
  return (
    <div className="card">
      <div className="card-header">
        <div className="row justify-content-between">
          <div>
            <span className="badge badge-primary badge-pill">{props.resource.subject}</span> 
            <span className="badge badge-secondary badge-pill">Grades: {displayGrade(props.resource.lowerGradeBound)}-{displayGrade(props.resource.upperGradeBound)}</span>
          </div>
          <div>
            {props.inPlaylist ? 
              <button className="non-button" onClick={() => props.removeResourceFromPlaylist(props.resource.id, props.playlist.id)}>
                <svg height="1rem" viewBox="-110 -18 581 581.33331" width="1rem" xmlns="http://www.w3.org/2000/svg"><path d="m181.25 363.109375 172.75 181.890625v-545h-360v545l172.75-181.890625c1.886719-1.984375 4.511719-3.113281 7.25-3.113281s5.359375 1.128906 7.25 3.113281zm0 0"/></svg>
              </button>
              :
              <Link to={{
                pathname: `/resources/playlists/new`,
                context: "resources", 
                state: {resourceId: props.resource.id}
              }}>
                <svg height="1rem" viewBox="-161 -27 859 859.3038" width="1rem" xmlns="http://www.w3.org/2000/svg"><path d="m514.824219 805.585938c-3.605469 0-7.140625-1.453126-9.726563-4.179688l-245.371094-258.277344-245.371093 258.277344c-3.792969 4-9.628907 5.273438-14.710938 3.226562-5.109375-2.03125-8.453125-6.972656-8.453125-12.46875v-765.3125c0-14.808593 12.046875-26.851562 26.855469-26.851562h483.351563c14.804687 0 26.855468 12.042969 26.855468 26.851562v765.308594c0 5.496094-3.34375 10.441406-8.441406 12.472656-1.617188.640626-3.308594.953126-4.988281.953126zm-255.097657-295.378907c3.671876 0 7.195313 1.507813 9.730469 4.179688l231.941407 244.144531v-731.679688h-483.351563v731.679688l231.941406-244.144531c2.535157-2.671875 6.058594-4.179688 9.738281-4.179688zm0 0"/></svg>
              </Link>
            }
          </div>
        </div>
      </div>
      <div className="card-body">
        <h4 classNmae="card-title">{props.resource.title}</h4>
      <br/>
        <p>{props.resource.description}</p>
        <Link className="btn btn-primary tertiary-background" to={`/resources/${props.resource.id}`}>See More</Link>
      </div>
    </div>
  )
}

export default ResourceThumbnail