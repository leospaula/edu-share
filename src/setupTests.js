// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
// import '@testing-library/jest-dom/extend-expect';

import React from 'react'
import { Router } from 'react-router-dom'
import { render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import rootReducer from './reducers/rootReducer'

function renderWithStore(
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}


function renderWithRouter(
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history,
  }
}

export { renderWithRouter, renderWithStore }

