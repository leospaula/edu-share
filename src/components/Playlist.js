import React from 'react'
import ResourceThumbnail from './ResourceThumbnail'

const Playlist = props => {
  if (!props.loadStatus || props.loadStatus === "pending"){
    return(
      <div>Loading...</div>
    )
  } else if (!props.playlist){
    return(
      <div>Resource Not Found</div>
    )
  } else {
    return(
      <div>
        <h2 className="my-3">{props.playlist.name}</h2>
        {props.playlist.resources.map(resource => 
          <div className="my-2">
            <ResourceThumbnail 
              key={resource.id} 
              resource={resource}
              inPlaylist={props.isResourceInPlaylist(resource.id)}
              playlist={props.playlist}
              removeResourceFromPlaylist={props.removeResourceFromPlaylist}
            />
          </div>
        )}
      </div>
    )
  }
}

export default Playlist