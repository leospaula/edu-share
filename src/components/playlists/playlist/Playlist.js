import React from 'react'
import ResourceThumbnail from '../../resources/ResourceThumbnail'

const Playlist = props => {
  if (!props.loadStatus || props.loadStatus === "pending"){
    return(
      <div data-testid="playlist-wrapper">Loading...</div>
    )
  } else if (!props.playlist){
    return(
      <div data-testid="playlist-wrapper">Resource Not Found</div>
    )
  } else {
    return(
      <div data-testid="playlist-wrapper">
        <h2 className="my-3">{props.playlist.name}</h2>
        {props.playlist.resources.map(resource => 
          <div data-testid="resource-wrapper" key={resource.id} className="my-2">
            <ResourceThumbnail 
              resource={resource}
              inPlaylist={props.isResourceInPlaylist(resource.id)}
              playlist={props.playlist}
              removeResourceFromPlaylist={props.removeResourceFromPlaylist}
            />
          </div>
        )}
      </div>
    )
  }
}

export default Playlist