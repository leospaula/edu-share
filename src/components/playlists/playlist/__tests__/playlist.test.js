import React from 'react'
import ReactDOM from 'react-dom'
import Playlist from '../Playlist'
import { BrowserRouter as Router } from 'react-router-dom'
import Adapter from 'enzyme-adapter-react-16'
import { shallow, configure } from 'enzyme'
import renderer from 'react-test-renderer'
import { cleanup } from '@testing-library/react'
import { renderWithRouter } from '../../../../setupTests'
import '@testing-library/jest-dom/extend-expect'
import ResourceThumbnail from '../../../resources/ResourceThumbnail'

configure({ adapter: new Adapter() });

const mockProps = {
  playlist: {
    id: 1,
    name: "Default Bookshelf",
    resources: [
      {
        title: "Photosynthesis | Educational Video for Kids",
        description: "Photosynthesis is a process used by plants and other organisms to convert light energy into chemical energy that can later be released to fuel the organisms' activities. This chemical energy is stored in carbohydrate molecules, such as sugars, which are synthesized from carbon dioxide and water – hence the name photosynthesis, from the Greek phōs (φῶς), light, and sunthesis (σύνθεσις), putting together.[1][2][3] In most cases, oxygen is also released as a waste product. Most plants, most algae, and cyanobacteria perform photosynthesis; such organisms are called photoautotrophs. Photosynthesis is largely responsible for producing and maintaining the oxygen content of the Earth's atmosphere, and supplies most of the energy necessary for life on Earth.",
        lowerGradeBound: 0,
        upperGradeBound: 5,
        url: "https://www.youtube.com/watch?v=UPBMG5EYydo",
        subject: "Biology",
        rating: 4,
        id: 1 
      },
      {
        title: "Instructions for Science Volcano",
        lowerGradeBound: 3,
        upperGradeBound: 8,
        url: "https://www.wikihow.com/Make-a-Volcano",
        subject: "Science",
        rating: 5,
        id: 2 
      }
    ],
    user_id: 1
  },
  loadStatus: null,
  isResourceInPlaylist: jest.fn().mockImplementation(() => true),
  removeResourceFromPlaylist: jest.fn()
}

afterEach(cleanup)

it ('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Playlist />, div)
})

it ('renders something regardless of whether or not a playlist gets loaded from the api', () => {
  const wrapper = shallow(<Playlist {...mockProps} />)
  expect(wrapper.exists('[data-testid="playlist-wrapper"]')).toBe(true)
  wrapper.setProps({ loadStatus: "pending" })
  expect(wrapper.exists('[data-testid="playlist-wrapper"]')).toBe(true)
  wrapper.setProps({ loadStatus: "complete" })
  expect(wrapper.exists('[data-testid="playlist-wrapper"]')).toBe(true)
  wrapper.setProps({ playlist: null })
  expect(wrapper.exists('[data-testid="playlist-wrapper"]')).toBe(true)
})

it ('matches the snapshot given mocked playlist data', () => {
  const tree = renderer.create(<Router>
      <Playlist {...mockProps} loadStatus={"complete"} />
    </Router>
    )
  expect(tree).toMatchSnapshot()
})

it ('displays the name of the playlist', () => {
  const { getByTestId } = renderWithRouter(<Playlist {...mockProps} loadStatus={"complete"}/>)
  expect(getByTestId('playlist-wrapper')).toHaveTextContent('Default Bookshelf')
})

it ('renders the correct number of resources per the playlist', () => {
  const { queryAllByTestId } = renderWithRouter(<Playlist {...mockProps} loadStatus={"complete"}/>)
  expect(queryAllByTestId('resource-wrapper')).toHaveLength(2)
})

it ('renders a ResourceThumbnail component per each resource in the playlist', () => {
  const wrapper = shallow(<Playlist {...mockProps} loadStatus={"complete"}/>)
  expect(wrapper.find(ResourceThumbnail)).toHaveLength(2)
})