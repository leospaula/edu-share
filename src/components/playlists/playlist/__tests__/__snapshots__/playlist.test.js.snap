// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot given mocked playlist data 1`] = `
<div
  data-testid="playlist-wrapper"
>
  <h2
    className="my-3"
  >
    Default Bookshelf
  </h2>
  <div
    className="my-2"
    data-testid="resource-wrapper"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        <div
          className="row justify-content-between"
        >
          <div>
            <span
              className="badge badge-primary badge-pill"
            >
              Biology
            </span>
            <span
              className="badge badge-secondary badge-pill"
            >
              Grades: 
              0
              -
              5
            </span>
          </div>
          <div>
            <button
              className="non-button"
              onClick={[Function]}
            >
              <svg
                height="1rem"
                viewBox="-110 -18 581 581.33331"
                width="1rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m181.25 363.109375 172.75 181.890625v-545h-360v545l172.75-181.890625c1.886719-1.984375 4.511719-3.113281 7.25-3.113281s5.359375 1.128906 7.25 3.113281zm0 0"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="card-body"
      >
        <h4
          className="card-title"
        >
          Photosynthesis | Educational Video for Kids
        </h4>
        <div
          className="embed-responsive embed-responsive-16by9"
        >
          <iframe
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
            allowFullScreen={true}
            className="embed-responsive-item"
            frameBorder="0"
            height="315"
            src="https://www.youtube.com/embed/UPBMG5EYydo"
            title="Photosynthesis | Educational Video for Kids"
            width="560"
          />
        </div>
        <p>
          Photosynthesis is a process used by plants and other organisms to convert light energy into chemical energy that can later be released to fuel the organisms' activities. This chemi...
        </p>
        <a
          className="btn btn-primary tertiary-background"
          href="/resources/1"
          onClick={[Function]}
        >
          See More
        </a>
      </div>
    </div>
  </div>
  <div
    className="my-2"
    data-testid="resource-wrapper"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        <div
          className="row justify-content-between"
        >
          <div>
            <span
              className="badge badge-primary badge-pill"
            >
              Science
            </span>
            <span
              className="badge badge-secondary badge-pill"
            >
              Grades: 
              3
              -
              8
            </span>
          </div>
          <div>
            <button
              className="non-button"
              onClick={[Function]}
            >
              <svg
                height="1rem"
                viewBox="-110 -18 581 581.33331"
                width="1rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m181.25 363.109375 172.75 181.890625v-545h-360v545l172.75-181.890625c1.886719-1.984375 4.511719-3.113281 7.25-3.113281s5.359375 1.128906 7.25 3.113281zm0 0"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="card-body"
      >
        <h4
          className="card-title"
        >
          Instructions for Science Volcano
        </h4>
        <p />
        <a
          className="btn btn-primary tertiary-background"
          href="/resources/2"
          onClick={[Function]}
        >
          See More
        </a>
      </div>
    </div>
  </div>
</div>
`;
